mesonbuild/mparser.py

```diff
@@ -213,42 +217,52 @@ def lex(self, subdir):
                 raise ParseException('lexer', self.getline(line_start), lineno, col)
 
 class BaseNode:
-    def accept(self, visitor):
+    def __init__(self, lineno: int, colno: int, subdir: str, end_lineno: Optional[int] = None, end_colno: Optional[int] = None) -> None:
+        self.lineno = lineno  # type: int
+        self.colno = colno    # type: int
+        self.subdir = subdir  # type: str
+        self.end_lineno = end_lineno if end_lineno is not None else self.lineno
+        self.end_colno = end_colno if end_colno is not None else self.colno
+
+        # Attributes for the visitors
+        self.level = 0            # type: int
+        self.ast_id = ''          # type: str
+        self.condition_level = 0  # type: int
+
+    def accept(self, visitor: 'AstVisitor') -> None:
         fname = 'visit_{}'.format(type(self).__name__)
         if hasattr(visitor, fname):
             func = getattr(visitor, fname)
             if callable(func):
                 func(self)
 
 class ElementaryNode(BaseNode):
-    def __init__(self, token):
-        self.lineno = token.lineno
-        self.subdir = token.subdir
-        self.colno = token.colno
+    def __init__(self, token: Token) -> None:
```

Comment:
I think then this would be

```python
def __init__(self, token: Token[TokenTypes]) -> None:
```

Reply:
Although actually this might need to be implemented in terms of `BaseNode` as a Generic... Not sure. Generics are kinda a bind bender.